[Stanley Tan] iP

stanley-1 opened the pull request:
## Duke v0.1
> "The roots of education are bitter, but the fruit is sweet." ~ Aristotle ([See more quotes](https://elementalscience.com/blogs/quotes-for-science/the-roots-of-education))

This primitive version of Duke helps you to keep track of your tasks! It is:
* Text-based
* Simple to use
* _VERY_ Responsive

To test this app,
1. Click [here](https://github.com/stanley-1/ip/releases)
2. Download the `ip.jar` file
3. Run `java -jar ip.jar` in your working directory
4. ~~**HAVE FUN!!**~~ :smile: 👍 

Features:

- [ ] Managing General Tasks
- [ ] Managing Deadlines with date formatting
- [ ] Managing Events
- [ ] Search/Filter tool
- [ ] GUI (coming soon!)

Sample chunk of code with fencing:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Thanks for reading this description and reviewing my codes! 😃

Commit:
Add Gradle support

Commit:
Added Level-1

Commit:
Added Level-2: Add, List

Commit:
Improved syntax by adding newlines to existing work.

Commit:
Created Task.java class and fulfilled Level 3

Commit:
Added toString() method for Task.java and implemented this with Duke.java

Commit:
Added ToDo.java which inherits from Task.java

Commit:
Added Deadline.java which inherits from Task.java

Commit:
Added Event.java which inherits from Task.java

Commit:
Modified runtest.sh for MacBook and crafted test scenarios.

Commit:
Created DukeException.java file & implemented Error Handling.

Commit:
Used ArrayList as Data Structure & implemented "Delete" functionality.

Commit:
Made preliminary edits to the README.

Commit:
Create static methods for Duke.java & edit EXPECTED.txt accordingly

Commit:
Add ability to read/write files

Commit:
Handle case where data file does not exist initially

Commit:
Accept date in yyyy-mm-dd and print in MMM dd yyyy format

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Handle different date formats

Commit:
Use more OOP in code

Commit:
Divide classes into packages

Commit:
Add 6 tests to DukeTest.java

Commit:
Add MANIFEST.MF to .gitignore

Commit:
Align format of code in Parser.java

Commit:
Add Find function in TaskList.java and Parser.java

Commit:
Tweak code to comply with Java coding standards

Commit:
Add JavaDoc comments

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	data/tasks.txt
#	src/main/java/TaskList.java

Commit:
Make minor edits to JavaDoc

Commit:
Implement minor edits

Commit:
Reverse errors in previous commit

Commit:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

Commit:
Add checkstyle.xml from AB-3 config files

Commit:
Merge branch 'branch-A-Checkstyle'

Commit:
Implement recommendations by Checkstyle plug-in

Commit:
Implement recommendations by Checkstyle plug-in

Commit:
Merge branch 'branch-self-move'

Chang-CH commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+
+                taskList.addDeadline(name + " /by " + when);
+                counter += 1;
+            }
```
Should use default else branch to catch unexpected characters.

Chang-CH commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,237 @@
+package duke;
+
+//import duke.Deadline;
```
No longer used  imports can be deleted instead.

Chang-CH commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,237 @@
+    /**
+     * Alerts user to an invalid command.
+     */
+    public void error() {
```
displayError might be a better name

Chang-CH commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+
+public class Storage {
+
+    static String dash = "__________________________________";
```
Dash is already defined and used in Ui, might be able to reuse code.

Chang-CH commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+            String toAdd = s.nextLine();
+
+            if (toAdd.charAt(1) == 'T') {
+                String task = toAdd.substring(7);
```
7 looks like a magic number, perhaps define it as a constant or add comments

Chang-CH commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,56 @@
+package duke;
```
hey, I like how clean this bit of code is 👍

Chang-CH reviewed (commented):
Overall, your code looks amazing, you followed the coding quality very closely. Some cosmetic changes were requested 👍

Commit:
Implement GUI for Duke

Commit:
Implement recommendations by Checkstyle

Commit:
Merge branch 'branch-Level-10'

seannzx commented on src/main/java/duke/TaskList.java:
```diff
@@ -1,18 +1,18 @@
-import duke.ToDo;
+//import duke.Deadline;
+//import duke.Event;
+//import duke.Storage;
```
As mentioned above, unused imports should be removed to improve readability

seannzx commented on src/main/java/Parser.java:
```diff
@@ -21,11 +21,11 @@ public void parseCommand(String userInput) {
             tasks.list();
 
-        }  else if (userInput.length() > 3 && userInput.substring(0, 4).matches("done")){
+        } else if (userInput.length() > 3 && userInput.substring(0, 4).matches("done")) {
```
Might want to consider utilizing `switch` statements to simplify long branches of `if-else` statements for readability

seannzx commented on src/main/java/duke/TaskList.java:
```diff
@@ -1,16 +1,9 @@
 
-    static String dash = "__________________________________";
     public ArrayList<Task> tasks;
     public int counter = 0;
```
Class fields should ideally be set to `private` in general.

seannzx commented on src/main/java/duke/Storage.java:
```diff
@@ -35,48 +34,56 @@ public Storage(String filepath, TaskList taskList) {
 
             } else if (toAdd.charAt(1) == 'E') {
                 String task = toAdd.substring(7);
                 String name = task.substring(0, task.indexOf("(") - 1);
```
Try to avoid duplicate lines of code. Consider helper functions to parse and extract similar functionalities.

reddevil1313 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,63 @@
+                storage.writeToFile(taskList.tasks);
+                return "Bye! Hope to see you again! :D";
+            } catch (IOException e) {
+                System.out.println("No tasks saved.");
```
Maybe you can use a better way to express this error.

reddevil1313 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,63 @@
+        } catch (FileNotFoundException e) {
+            System.out.println("File not found");
+        }
+        return "You don't have any saved tasks";
```
Same as before, I feel the error should identify what the issue is to the user and how he /she can rectify it.

reddevil1313 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
What's the reason for making this field protected?

reddevil1313 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,47 @@
+package duke;
```
This file is very clean and well documented!

reddevil1313 commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -5,3 +5,51 @@ Hello from
 |____/ \__,_|_|\_\___|
 
+__________________________________
+Howdy! I'm Duke
```
Maybe you can have more different cases for testing? Like deleting a task!

reddevil1313 reviewed (commented):
Good Job! Your code looks quite clean and very well documented! I couldn't find any errors except for some very minute details!

Commit:
Improve code readability

Presence of superfluous import statements.

Various classes include unnecessary import statements, resulting in
cluttered code and poor readability.

Let's delete unused import statements that have already been commented
out from the existing codebase to reduce unnecessary noise and improve
code readability.

Commit:
Teach Duke to deal with empty TaskList

Duke prints the incorrect message when TaskList is empty.

Printing the same message regardless of TaskList's size is erroneous
and might confuse users.

Let's update list() method to reflect "You have no tasks in your list!"
message when the user's TaskList is empty.

Making this change improves the accuracy of results produced by Duke.

Commit:
Implement Assertions within Storage.java and TaskList.java

Storage and TaskList classes lack an internal check to ensure that
the counter is larger than zero (i.e.: TaskList is not empty).

Let's implement Java Assert statements at critical junctures in methods
belonging to Storage and TaskList classes.

Using Java Assert statements allows us to detect bugs in our program
and rectify them before they are discovered by the end-user.

Commit:
Improve general quality of codebase

Variables and attributes were not marked as "private" or "final" during
initial coding for convenience in testing and experimentation.

Let's declare all necessary variables as "private" and "final". At the
same time, create getters to extract relevant details from such
variables.

Declaring majority of variables as "private" and "final" satisfies the
criteria for immutability under Java Coding Standards.

Commit:
Delete legacy class that supports Text-Based UI

Commit:
Delete legacy code used in Text-Based UI

Commit:
Replace magic numbers with named constants

Commit:
Refactor repeated code into an auxiliary method

Three methods (makeTodo, makeDeadline, makeEvent) exhibit common
behavior.

The common behavior across these three methods results in code
duplication.

Extracting the common behavior into an auxiliary method (makeTask())
allows us to avoid code duplication and the presence of long methods.

Let's extract behavior common to these Task Creation methods into a new
auxiliary method named makeTask().

stanley-1 committed:
Merge pull request #1 from stanley-1/branch-A-Assertions

Implement Assertions within Storage.java and TaskList.java

Commit:
Merge branch 'master' of https://github.com/stanley-1/ip

* 'master' of https://github.com/stanley-1/ip:
  Implement Assertions within Storage.java and TaskList.java

Commit:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Refactor repeated code into an auxiliary method
  Replace magic numbers with named constants
  Delete legacy code used in Text-Based UI
  Delete legacy class that supports Text-Based UI
  Improve general quality of codebase

# Conflicts:
#	src/main/java/duke/TaskList.java

Solved by deleting legacy code.

Commit:
Implement in-App guidance for users

Commit:
Merge branch 'branch-C-Help'

* branch-C-Help:
  Implement in-App guidance for users

Commit:
Minor syntax improvement

Commit:
Implement Error Handling via DukeException

Commit:
Implement recommendation by cs2103-bot

Commit:
Improve the GUI

Commit:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Improve the GUI

Commit:
Upload product screenshot

Commit:
Update README.md

Commit:
Update to User Guide

Commit:
Improve Help tool
